pkg/gpu/stream.go

```diff
@@ -67,17 +80,68 @@ type kernelSpan struct {
 
 	// numKernels is the number of kernels that were launched during the span
 	numKernels uint64
+
+	// avgMemoryUsage is the average memory usage during the span, per allocation type
+	avgMemoryUsage map[model.MemAllocType]uint64
```

Comment:
why do we need to calculate those here? can't it be calculated at the backend/frontend level?
I think most of our avg/max type metrics are not calculated on the agent side (worth double checking)

Reply:
It's the average for this span, it's calculated based on all the kernel launches pertaining to it. We could send the total memory usage and then divide in the backend but I think it makes more sense to make that simple calculation here to avoid sending more data.

Reply:
but we send the total memory anyway. i haven't found avg memory metric for host memory, do you have an example?

Reply:
The total memory we send in the core check is not exactly the same metric we are calculating here. Here we have a sequence of kernels, each one with their own associated memory usage due to kernel/shared/global memory. We cannot add all that memory and send it as the total, as they're not executing concurrently. We calculate here how much memory, on average, has this span consumed over its lifetime.

Reply:
and that metric was requested? i prefer to add those extra metrics upon request in the future and keep the code as simple as possible

Reply:
It's not a new metric, just extra data to have the `gpu.memory` metric to be as close to the results we got from the NVIDIA profiler.

Reply:
we will discuss it separately as well. currently not a blocker for this PR